File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/SfapApiClient.kt
/*
 * Copyright (c) 2024-present, salesforce.com, inc.
 * All rights reserved.
 * Redistribution and use of this software in source and binary forms, with or
 * without modification, are permitted provided that the following conditions
 * are met:
 * - Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * - Neither the name of salesforce.com, inc. nor the names of its contributors
 * may be used to endorse or promote products derived from this software without
 * specific prior written permission of salesforce.com, inc.
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package com.salesforce.androidsdk.rest

import com.salesforce.androidsdk.rest.RestRequest.RestMethod.POST
import com.salesforce.androidsdk.rest.SfapApiGenerationsResponseBody.Companion.fromJson
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody.Companion.toRequestBody

/**
 * Provides REST client methods for a variety of `sfap_api` endpoints.
 * - `chat-generations`
 * - `embeddings`
 * - `feedback`
 * - `generations`
 *
 * See https://developer.salesforce.com/docs/einstein/genai/guide/access-models-api-with-rest.html
 *
 * @param apiHostName The Salesforce `sfap_api` hostname
 * @param modelName The model name to request from.  For possible
 * values, see https://developer.salesforce.com/docs/einstein/genai/guide/api-names.html.
 * Note that the `embeddings` endpoint requires an embeddings-enabled model such
 * as `sfdc_ai__DefaultOpenAITextEmbeddingAda_002`.  Also note that submitting
 * to the `feedback` endpoint does not require value for this parameter
 * @param restClient The REST client to use
 */
class SfapApiClient(
    private val apiHostName: String,
    private val modelName: String? = null,
    private val restClient: RestClient
) {

    /**
     * Submit a request to the `sfap_api` `embeddings` endpoint.
     * @param requestBody The `embeddings` request body
     * @return The endpoint's response
     */
    @Throws(SfapApiException::class)
    fun fetchGeneratedEmbeddings(
        requestBody: SfapApiEmbeddingsRequestBody
    ): SfapApiEmbeddingsResponseBody {

        // Submit the request.
        val restRequest = RestRequest(
            POST,
            "https://$apiHostName/einstein/platform/v1/models/$modelName/embeddings",
            requestBody
                .toJson()
                .toRequestBody(CONTENT_TYPE_HEADER_VALUE_APPLICATION_JSON.toMediaTypeOrNull()),
            generateSfapApiHeaders()
        )
        val restResponse = restClient.sendSync(restRequest)
        val responseBodyString = restResponse.asString()
        return if (restResponse.isSuccess && responseBodyString != null) {
            SfapApiEmbeddingsResponseBody.fromJson(responseBodyString)
        } else {
            val errorResponseBody = SfapApiErrorResponseBody.fromJson(responseBodyString)
            throw SfapApiException(
                errorCode = errorResponseBody.errorCode,
                message = responseBodyString,
                messageCode = errorResponseBody.messageCode,
                source = responseBodyString
            )
        }
    }

    /**
     * Fetches generated chat responses from the `sfap_api` `chat-generations`
     * endpoint.
     * @param requestBody The `chat-generations` request body
     * @return The endpoint's response
     */
    @Throws(SfapApiException::class)
    fun fetchGeneratedChat(
        requestBody: SfapApiChatGenerationsRequestBody
    ): SfapApiChatGenerationsResponseBody {

        // Request the generated chat responses.
        val restRequest = RestRequest(
            POST,
            "https://$apiHostName/einstein/platform/v1/models/$modelName/chat-generations",
            requestBody.toJson()
                .toRequestBody(CONTENT_TYPE_HEADER_VALUE_APPLICATION_JSON.toMediaTypeOrNull()),
            generateSfapApiHeaders()
        )
        val restResponse = restClient.sendSync(restRequest)

        val responseBodyString = restResponse.asString()
        return if (restResponse.isSuccess && responseBodyString != null) {
            SfapApiChatGenerationsResponseBody.fromJson(responseBodyString)
        } else {
            val errorResponseBody = SfapApiErrorResponseBody.fromJson(responseBodyString)
            throw SfapApiException(
                errorCode = errorResponseBody.errorCode,
                message = responseBodyString,
                messageCode = errorResponseBody.messageCode,
                source = responseBodyString
            )
        }
    }

    /**
     * Fetches generated text from the `sfap_api` "generations" endpoint.
     * @param prompt The prompt request parameter value
     * @return The endpoint's response
     */
    @Throws(SfapApiException::class)
    fun fetchGeneratedText(
        prompt: String
    ): SfapApiGenerationsResponseBody {

        // Request the generated text.
        val restRequest = RestRequest(
            POST,
            "https://$apiHostName/einstein/platform/v1/models/$modelName/generations",
            SfapApiGenerationsRequestBody(prompt)
                .toJson()
                .toRequestBody(CONTENT_TYPE_HEADER_VALUE_APPLICATION_JSON.toMediaTypeOrNull()),
            generateSfapApiHeaders()
        )
        val restResponse = restClient.sendSync(restRequest)

        val responseBodyString = restResponse.asString()
        return if (restResponse.isSuccess && responseBodyString != null) {
            fromJson(responseBodyString)
        } else {
            val errorResponseBody = SfapApiErrorResponseBody.fromJson(responseBodyString)
            throw SfapApiException(
                errorCode = errorResponseBody.errorCode,
                message = responseBodyString,
                messageCode = errorResponseBody.messageCode,
                source = responseBodyString
            )
        }
    }

    /**
     * Submits feedback for previously generated text from the `sfap_api`
     * endpoints to the `sfap_api` `feedback` endpoint.
     * @param requestBody The `feedback` request body
     * @return The endpoint's response
     */
    @Throws(SfapApiException::class)
    fun submitGeneratedTextFeedback(
        requestBody: SfapApiFeedbackRequestBody
    ): SfapApiFeedbackResponseBody {

        // Submit the feedback.
        val restRequest = RestRequest(
            POST,
            "https://$apiHostName/einstein/platform/v1/feedback",
            requestBody
                .toJson()
                .toRequestBody(CONTENT_TYPE_HEADER_VALUE_APPLICATION_JSON.toMediaTypeOrNull()),
            generateSfapApiHeaders()
        )
        val restResponse = restClient.sendSync(restRequest)

        val responseBodyString = restResponse.asString()
        return if (restResponse.isSuccess && responseBodyString != null) {
            SfapApiFeedbackResponseBody.fromJson(responseBodyString)
        } else {
            val errorResponseBody = SfapApiErrorResponseBody.fromJson(responseBodyString)
            throw SfapApiException(
                errorCode = errorResponseBody.errorCode,
                message = responseBodyString,
                messageCode = errorResponseBody.messageCode,
                source = responseBodyString
            )
        }
    }

    companion object {

        internal val jsonIgnoreUnknownKeys = Json { ignoreUnknownKeys = true }

        private const val CONTENT_TYPE_HEADER_VALUE_APPLICATION_JSON = "application/json; charset=utf-8"

        private fun generateSfapApiHeaders() = mutableMapOf(
            "x-sfdc-app-context" to "EinsteinGPT",
            "x-client-feature-id" to "ai-platform-models-connected-app"
        )
    }
}

File: flutter_sdk/lib/src/core/rest/sfap_api_client.dart
import 'dart:convert';
import 'rest_client.dart';
import 'rest_request.dart';
import 'rest_response.dart';

/// Exception for SFAP (Einstein Platform) API errors.
class SfapApiException implements Exception {
  final String? errorCode;
  final String? message;
  final String? messageCode;
  final String? source;

  SfapApiException({
    this.errorCode,
    this.message,
    this.messageCode,
    this.source,
  });

  factory SfapApiException.fromResponse(RestResponse response) {
    try {
      final json = jsonDecode(response.asString) as Map<String, dynamic>;
      return SfapApiException(
        errorCode: json['errorCode'] as String?,
        message: json['message'] as String?,
        messageCode: json['messageCode'] as String?,
        source: response.asString,
      );
    } catch (_) {
      return SfapApiException(
        message: 'HTTP ${response.statusCode}',
        source: response.asString,
      );
    }
  }

  @override
  String toString() => 'SfapApiException: $message ($errorCode)';
}

/// Chat message for SFAP chat generations.
class SfapChatMessage {
  final String role;
  final String content;

  const SfapChatMessage({required this.role, required this.content});

  Map<String, dynamic> toJson() => {'role': role, 'content': content};

  factory SfapChatMessage.fromJson(Map<String, dynamic> json) {
    return SfapChatMessage(
      role: json['role'] as String,
      content: json['content'] as String,
    );
  }
}

/// Request body for embeddings generation.
class SfapEmbeddingsRequestBody {
  final List<String> input;

  const SfapEmbeddingsRequestBody({required this.input});

  Map<String, dynamic> toJson() => {'input': input};
}

/// Response from embeddings generation.
class SfapEmbeddingsResponseBody {
  final List<Map<String, dynamic>>? data;
  final Map<String, dynamic>? usage;

  SfapEmbeddingsResponseBody({this.data, this.usage});

  factory SfapEmbeddingsResponseBody.fromJson(Map<String, dynamic> json) {
    return SfapEmbeddingsResponseBody(
      data: (json['data'] as List<dynamic>?)
          ?.cast<Map<String, dynamic>>(),
      usage: json['usage'] as Map<String, dynamic>?,
    );
  }
}

/// Request body for chat generations.
class SfapChatGenerationsRequestBody {
  final List<SfapChatMessage> messages;
  final Map<String, dynamic>? parameters;

  const SfapChatGenerationsRequestBody({
    required this.messages,
    this.parameters,
  });

  Map<String, dynamic> toJson() => {
        'messages': messages.map((m) => m.toJson()).toList(),
        if (parameters != null) ...parameters!,
      };
}

/// Response from chat generations.
class SfapChatGenerationsResponseBody {
  final String? id;
  final String? generationDetails;
  final Map<String, dynamic>? raw;

  SfapChatGenerationsResponseBody({
    this.id,
    this.generationDetails,
    this.raw,
  });

  factory SfapChatGenerationsResponseBody.fromJson(
      Map<String, dynamic> json) {
    return SfapChatGenerationsResponseBody(
      id: json['id'] as String?,
      generationDetails: json['generationDetails'] as String?,
      raw: json,
    );
  }
}

/// Response from text generations.
class SfapGenerationsResponseBody {
  final String? id;
  final Map<String, dynamic>? generation;
  final Map<String, dynamic>? raw;

  SfapGenerationsResponseBody({this.id, this.generation, this.raw});

  factory SfapGenerationsResponseBody.fromJson(Map<String, dynamic> json) {
    return SfapGenerationsResponseBody(
      id: json['id'] as String?,
      generation: json['generation'] as Map<String, dynamic>?,
      raw: json,
    );
  }
}

/// Request body for feedback submission.
class SfapFeedbackRequestBody {
  final String generationId;
  final String feedback;
  final String? feedbackText;
  final String? source;
  final String? appFeedback;

  const SfapFeedbackRequestBody({
    required this.generationId,
    required this.feedback,
    this.feedbackText,
    this.source,
    this.appFeedback,
  });

  Map<String, dynamic> toJson() => {
        'generationId': generationId,
        'feedback': feedback,
        if (feedbackText != null) 'feedbackText': feedbackText,
        if (source != null) 'source': source,
        if (appFeedback != null) 'appFeedback': appFeedback,
      };
}

/// Response from feedback submission.
class SfapFeedbackResponseBody {
  final String? id;
  final Map<String, dynamic>? raw;

  SfapFeedbackResponseBody({this.id, this.raw});

  factory SfapFeedbackResponseBody.fromJson(Map<String, dynamic> json) {
    return SfapFeedbackResponseBody(
      id: json['id'] as String?,
      raw: json,
    );
  }
}

/// REST client for Salesforce Einstein Platform (SFAP) API.
///
/// Supports chat generations, text generations, embeddings, and feedback.
/// Mirrors Android SfapApiClient for feature parity.
class SfapApiClient {
  final String apiHostName;
  final String? modelName;
  final RestClient _restClient;

  static const String _basePath = '/einstein/platform/v1';

  SfapApiClient({
    required this.apiHostName,
    this.modelName,
    required RestClient restClient,
  }) : _restClient = restClient;

  /// Standard SFAP API headers.
  static Map<String, String> generateHeaders() => {
        'x-sfdc-app-context': 'EinsteinGPT',
        'x-client-feature-id': 'ai-platform-models-connected-app',
      };

  /// Generates embeddings for the given input texts.
  ///
  /// Requires [modelName] to be set.
  Future<SfapEmbeddingsResponseBody> fetchGeneratedEmbeddings(
      SfapEmbeddingsRequestBody requestBody) async {
    _requireModelName();
    final request = _buildRequest(
      '/models/$modelName/embeddings',
      requestBody.toJson(),
    );
    final response = await _restClient.sendAsync(request);
    if (!response.isSuccess) throw SfapApiException.fromResponse(response);
    return SfapEmbeddingsResponseBody.fromJson(response.asJsonObject());
  }

  /// Generates chat completions.
  ///
  /// Requires [modelName] to be set.
  Future<SfapChatGenerationsResponseBody> fetchGeneratedChat(
      SfapChatGenerationsRequestBody requestBody) async {
    _requireModelName();
    final request = _buildRequest(
      '/models/$modelName/chat-generations',
      requestBody.toJson(),
    );
    final response = await _restClient.sendAsync(request);
    if (!response.isSuccess) throw SfapApiException.fromResponse(response);
    return SfapChatGenerationsResponseBody.fromJson(response.asJsonObject());
  }

  /// Generates text from a prompt.
  ///
  /// Requires [modelName] to be set.
  Future<SfapGenerationsResponseBody> fetchGeneratedText(String prompt) async {
    _requireModelName();
    final request = _buildRequest(
      '/models/$modelName/generations',
      {'prompt': prompt},
    );
    final response = await _restClient.sendAsync(request);
    if (!response.isSuccess) throw SfapApiException.fromResponse(response);
    return SfapGenerationsResponseBody.fromJson(response.asJsonObject());
  }

  /// Submits feedback for a generation.
  Future<SfapFeedbackResponseBody> submitFeedback(
      SfapFeedbackRequestBody requestBody) async {
    final request = _buildRequest('/feedback', requestBody.toJson());
    final response = await _restClient.sendAsync(request);
    if (!response.isSuccess) throw SfapApiException.fromResponse(response);
    return SfapFeedbackResponseBody.fromJson(response.asJsonObject());
  }

  RestRequest _buildRequest(String pathSuffix, Map<String, dynamic> body) {
    return RestRequest(
      method: RestMethod.post,
      path: 'https://$apiHostName$_basePath$pathSuffix',
      requestBody: body,
      additionalHttpHeaders: generateHeaders(),
    );
  }

  void _requireModelName() {
    if (modelName == null || modelName!.isEmpty) {
      throw StateError('modelName is required for this SFAP API call');
    }
  }
}
